L1Trigger/Phase2L1ParticleFlow/plugins/L1TPFCaloProducer.cc

```diff
@@ -0,0 +1,296 @@
+        hcalClusterer_.add(it);
+      }
+    }
+    if (!hcalDigis_.empty()) {
+      readHcalDigis_(iEvent, iSetup);
+    }
+    if (!phase2barrelTowers_.empty()) {
+      readPhase2BarrelCaloTowers_(iEvent, iSetup);
+    }
+    if (!hcalHGCTowers_.empty()) {
+      readHcalHGCTowers_(iEvent, iSetup);
+    }
+  }
+
+  /// --------------- CLUSTERING ------------------
+  ecalClusterer_.run();
+
+  auto ecalCellsH = iEvent.put(ecalClusterer_.fetchCells(), "ecalCells");
+
+  iEvent.put(ecalClusterer_.fetch(ecalCellsH), "emUncalibrated");
+
+  if (emCorrector_.valid()) {
+    ecalClusterer_.correct(
+        [&](const l1tpf_calo::Cluster &c) -> float { return emCorrector_.correctedPt(0., c.et, std::abs(c.eta)); });
+  }
+
+  std::unique_ptr<l1t::PFClusterCollection> corrEcal = ecalClusterer_.fetch(ecalCellsH);
+
+  if (debug_) {
+    for (const l1t::PFCluster &it : *corrEcal) {
+      std::cout << "L1TPFCaloProducer: corrected ECal cluster pt " << it.pt() << ", eta " << it.eta() << ", phi "
+                << it.phi() << std::endl;
+    }
+  }
+
+  auto ecalClustH = iEvent.put(std::move(corrEcal), "emCalibrated");
+
+  if (ecalOnly_) {
+    ecalClusterer_.clear();
+    return;
+  }
+
+  hcalClusterer_.run();
+
+  auto hcalCellsH = iEvent.put(hcalClusterer_.fetchCells(), "hcalCells");
+
+  // this we put separately for debugging
+  iEvent.put(hcalClusterer_.fetchCells(/*unclustered=*/true), "hcalUnclustered");
+
+  iEvent.put(hcalClusterer_.fetch(hcalCellsH), "hcalUncalibrated");
+
+  if (hcCorrector_.valid()) {
+    hcalClusterer_.correct(
+        [&](const l1tpf_calo::Cluster &c) -> float { return hcCorrector_.correctedPt(c.et, 0., std::abs(c.eta)); });
+  }
+
+  auto hcalClustH = iEvent.put(hcalClusterer_.fetch(hcalCellsH), "hcalCalibrated");
+
+  // Calorimeter linking
+  caloLinker_->run();
+
+  iEvent.put(caloLinker_->fetch(ecalClustH, hcalClustH), "uncalibrated");
+
+  if (hadCorrector_.valid()) {
+    caloLinker_->correct([&](const l1tpf_calo::CombinedCluster &c) -> float {
+      if (debug_)
+        std::cout << "L1TPFCaloProducer: raw linked cluster pt " << c.et << ", eta " << c.eta << ", phi " << c.phi
+                  << ", emPt " << c.ecal_et << std::endl;
+      return hadCorrector_.correctedPt(c.et, c.ecal_et, std::abs(c.eta));
+    });
+  }
+
+  std::unique_ptr<l1t::PFClusterCollection> clusters = caloLinker_->fetch(ecalClustH, hcalClustH);
+  for (l1t::PFCluster &c : *clusters) {
+    c.setPtError(resol_(c.pt(), std::abs(c.eta())));
+    if (debug_)
+      std::cout << "L1TPFCaloProducer: calibrated linked cluster pt " << c.pt() << ", eta " << c.eta() << ", phi "
+                << c.phi() << ", emPt " << c.emEt() << std::endl;
+  }
+  iEvent.put(std::move(clusters), "calibrated");
+
+  ecalClusterer_.clear();
+  hcalClusterer_.clear();
+  caloLinker_->clear();
+}
+
+void L1TPFCaloProducer::readHcalDigis_(edm::Event &iEvent, const edm::EventSetup &iSetup) {
+  iSetup.get<CaloTPGRecord>().get(decoder_);
```

Comment:
Please consider using `ESGetToken` https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideHowToGetDataFromES#In_ED_module. This needs to be changed anyway soon.

Reply:
Ok, done